src/builtin/ml.ts

```diff
@@ -0,0 +1,225 @@
+import * as Sphinx from 'sphinx-bot'
+import { finalAction } from '../controllers/botapi'
+import { ChatRecord, models, ChatBotRecord } from '../models'
+import { findBot, botResponse } from './utill'
+import { loadConfig } from '../utils/config'
+import fetch from 'node-fetch'
+
+const config = loadConfig()
+
+const msg_types = Sphinx.MSG_TYPE
+
+let initted = false
+
+export const ML_PREFIX = '/ml'
+
+export const ML_BOTNAME = `${ML_PREFIX.substring(
+  1,
+  2
+).toUpperCase()}${ML_PREFIX.substring(2)}Bot`
+
+type ContentKind = 'text' | 'image'
+
+export const CALLBACKS: { [k: string]: (msg: string) => void } = {}
+
+export function init() {
+  if (initted) return
+  initted = true
+
+  const client = new Sphinx.Client()
+  client.login('_', finalAction)
+
+  client.on(msg_types.MESSAGE, async (message: Sphinx.Message) => {
+    if (message.author?.bot !== ML_PREFIX) return
+    const isAdmin = message.member.roles.find((role) => role.name === 'Admin')
+    try {
+      const tribe = (await models.Chat.findOne({
+        where: { uuid: message.channel.id },
+      })) as ChatRecord
+
+      const bot: ChatBotRecord = await findBot({ botPrefix: ML_PREFIX, tribe })
+
+      let meta: MlMeta = JSON.parse(bot.meta || `{}`)
+      meta.kind = meta.kind || 'text'
+      const url = meta.url
+      const api_key = meta.apiKey
+      const arr = (message.content && message.content.split(' ')) || []
+
+      if (isAdmin && arr[0] === ML_PREFIX) {
+        const cmd = arr[1]
+
+        switch (cmd) {
+          case 'url':
+            const newUrl = arr[2]
+            if (!newUrl) {
+              await botResponse(
+                ML_BOTNAME,
+                'Please provide a valid URL',
+                ML_PREFIX,
+                tribe.id,
+                message,
+                cmd
+              )
+              return
+            }
+            meta.url = newUrl
+            await bot.update({ meta: JSON.stringify(meta) })
+            await botResponse(
+              ML_BOTNAME,
+              'URL updated successfully',
+              ML_PREFIX,
+              tribe.id,
+              message,
+              cmd
+            )
+            return
```

Comment:
for each of these switch statements it would be much cleaner if we just had a function for each one
like
```
case 'url':
   await url()
case 'api_key':
   await api_key()
case 'kind':
   await kind()
default:
   defaultFunc()
```